JSON string with bool value conversion issue

djjsdtc opened the pull request:
See issue #2091

domaindrivendev commented:
Good catch - this is definitely a bug, and there could be other similar ones due to the use of `ToString()`. Regarding the fix, I think we can reduce the amount of duplicated logic, and get away from `ToString()` altogether by taking a slightly different approach ...

I think we can extract a new _private_ method `CreateFromJsonElement(JsonElement)` out of `CreateFromJson(string)`. For example, the latter would be reduced to the following:

```csharp
public static IOpenApiAny CreateFromJson(string json)
{
    try
    {
        var jsonElement = JsonSerializer.Deserialize<JsonElement>(json);

        return CreateFromJsonElement(jsonElement);
    }
    catch { }

    return null;
}
```

And, all of the `JsonElement` to `IOpenApiAny` mapping code would be moved into `CreateFromJsonElement(JsonElement)`. Then finally, the code in the object and array specific mapping methods could just call the new method. For example, `CreateOpenApiArray` could be simplified to the following:

```csharp
private static IOpenApiAny CreateOpenApiArray(JsonElement jsonElement)
{
    var openApiArray = new OpenApiArray();

    foreach (var item in jsonElement.EnumerateArray())
    {
        openApiArray.Add(CreateFromJsonElement(item));
    }

    return openApiArray;
}
```

I believe this will solve the issue and make the code a lot more resilient. Additionally, this PR should be accompanied by at least one test. You could probably just add an additional case to the following:
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/master/test/Swashbuckle.AspNetCore.SwaggerGen.Test/SchemaGenerator/JsonSerializerSchemaGeneratorTests.cs#L267

djjsdtc commented:
> Good catch - this is definitely a bug, and there could be other similar ones due to the use of `ToString()`. Regarding the fix, I think we can reduce the amount of duplicated logic, and get away from `ToString()` altogether by taking a slightly different approach ...
> 
> I think we can extract a new _private_ method `CreateFromJsonElement(JsonElement)` out of `CreateFromJson(string)`. So, the latter would be reduced to the following:
> 
> ```cs
> public static IOpenApiAny CreateFromJson(string json)
> {
>     try
>     {
>         var jsonElement = JsonSerializer.Deserialize<JsonElement>(json);
> 
>         return CreateFromJsonElement(jsonElement);
>     }
>     catch { }
> 
>     return null;
> }
> ```
> 
> All of the `JsonElement` to `IOpenApiAny` mapping code would be moved into `CreateFromJsonElement(JsonElement)`. Then finally, the code in the object and array specific mapping methods could just call the new method. For example, `CreateOpenApiArray` could be simplified to the following:
> 
> ```cs
> private static IOpenApiAny CreateOpenApiArray(JsonElement jsonElement)
> {
>     var openApiArray = new OpenApiArray();
> 
>     foreach (var item in jsonElement.EnumerateArray())
>     {
>         openApiArray.Add(CreateFromJsonElement(item));
>     }
> 
>     return openApiArray;
> }
> ```
> 
> I believe this will solve the issue and make the code a lot more resilient. Additionally, this PR should be accompanied by at least one test. You could probably just add an additional case to the following:
> https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/master/test/Swashbuckle.AspNetCore.SwaggerGen.Test/SchemaGenerator/JsonSerializerSchemaGeneratorTests.cs#L267

Thanks for suggestion. I have moved the code outside to another method, and added unit tests for this method.

domaindrivendev commented:
This is great - thankyou 👍
